File: merge_sort.py
import csv
import sys
import time

def read_csv(filename):
    with open(filename, newline='') as f:
        return [[int(row[0]), row[1]] for row in csv.reader(f)]

def write_csv(data, filename):
    with open(filename, "w", newline='') as f:
        writer = csv.writer(f)
        writer.writerows(data)

def merge_sort(arr, left, right):
    if left < right:
        mid = (left + right) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)

def merge(arr, left, mid, right):
    L = arr[left:mid + 1]
    R = arr[mid + 1:right + 1]
    i = j = 0
    k = left
    while i < len(L) and j < len(R):
        if L[i][0] <= R[j][0]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
    while i < len(L):
        arr[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        arr[k] = R[j]
        j += 1
        k += 1

def main():
    if len(sys.argv) != 2:
        print("Usage: python merge_sort.py <dataset.csv>")
        return
    filename = sys.argv[1]
    data = read_csv(filename)
    start = time.time()
    merge_sort(data, 0, len(data) - 1)
    end = time.time()
    output = f"merge_sort_{len(data)}.csv"
    write_csv(data, output)
    print(f"Sorted file saved to {output}")
    print(f"Execution time: {end - start:.3f} seconds")

if __name__ == "__main__":
    main()

File: merge_sort.java
package ass;

import java.io.*;
import java.util.*;

public class merge_sort {
    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.out.println("Usage: java Ass.merge_sort <dataset_file>");
            return;
        }

        String filename = args[0];
        List<DataPair> data = readCSV(filename);

        long startTime = System.nanoTime();
        mergeSort(data, 0, data.size() - 1);
        long endTime = System.nanoTime();

        String outFile = "merge_sort_" + data.size() + ".csv";
        writeCSV(data, outFile);
        System.out.println("Sorted file saved to: " + outFile);
        System.out.printf("Execution time: %.3f seconds\n", (endTime - startTime) / 1e9);
    }

    static List<DataPair> readCSV(String filename) throws IOException {
        List<DataPair> list = new ArrayList<>();
        BufferedReader reader = new BufferedReader(new FileReader(filename));
        String line;
        while ((line = reader.readLine()) != null) {
            String[] parts = line.split(",");
            list.add(new DataPair(Integer.parseInt(parts[0]), parts[1]));
        }
        reader.close();
        return list;
    }

    static void mergeSort(List<DataPair> list, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(list, left, mid);
            mergeSort(list, mid + 1, right);
            merge(list, left, mid, right);
        }
    }

    static void merge(List<DataPair> list, int left, int mid, int right) {
        LinkedList<DataPair> L = new LinkedList<>(list.subList(left, mid + 1));
        LinkedList<DataPair> R = new LinkedList<>(list.subList(mid + 1, right + 1));
        int k = left;
        while (!L.isEmpty() && !R.isEmpty()) {
            if (L.peek().number <= R.peek().number) {
                list.set(k++, L.poll());
            } else {
                list.set(k++, R.poll());
            }
        }
        while (!L.isEmpty()) list.set(k++, L.poll());
        while (!R.isEmpty()) list.set(k++, R.poll());
    }

    static void writeCSV(List<DataPair> list, String filename) throws IOException {
        BufferedWriter writer = new BufferedWriter(new FileWriter(filename));
        for (DataPair dp : list) {
            writer.write(dp.number + "," + dp.text);
            writer.newLine();
        }
        writer.close();
    }

    static class DataPair {
        int number;
        String text;

        public DataPair(int number, String text) {
            this.number = number;
            this.text = text;
        }
    }
}